metaswitch/common/alarms.py

```diff
@@ -31,36 +31,281 @@
+
+        If the alarm described by alarm_handle has a single non-cleared severity,
+        return an Alarm representing the object. If the alarm has multiple
+        severities, return a MultiSeverityAlarm.
+
+        alarm_handle should be a constant defined in alarm_constants.
+        alarm_constants is generated for a project by running alarm_writer.py
+        over the JSON alarm definition file. For all of our projects, this is
+        done as part of the build process.
+
+        Alarm handles should be of the following form:
+        `(<index_number>, <severity1>, <severity2>, ...)`
+        """
+        # Prevent two threads from creating the same alarm object.
+        with self._registry_lock:
+            alarm = self._alarm_registry.get((issuer, alarm_handle), None)
+
+            if not alarm:
+                # See format description in docstring.
+                index = alarm_handle[0]
+                severities = list(alarm_handle[1:])
+                severities.remove(CLEARED)
+
+                if len(severities) == 1:
+                    alarm = Alarm(issuer, index, severities[0])
+                elif len(severities) > 1:
+                    alarm = MultiSeverityAlarm(issuer, index, severities)
+                else:
+                    raise ValueError('alarm_handle must contain a severity.')
+
+                self._alarm_registry[(issuer, alarm_handle)] = alarm
+                should_start = (not self._running) and (not self._should_terminate)
+
+        # It would be wasteful to start the alarm re-sync thread with
+        # no alarms present.
+        if should_start:
+            self.start()
+
+            # Note, this creates a reference to the alarm_manager, which
+            # means that it will not be cleaned up until exit. This is
+            # fine for a singleton which is expected to run until exit.
+            atexit.register(self.terminate)
+
+        return alarm
+
+    def run(self):
+        """Run loop to keep alarms in sync."""
+        with self._condition:
+            while True:
+                sleep_time = self._update_resync_time()
+
+                # Cope with the fact that we may be woken up early and
+                # have to sleep again.
+                while (sleep_time > 0):
+                    self._condition.wait(sleep_time)
+                    if self._should_terminate:
+                        break
+                    sleep_time = self._next_resync_time - monotonic()
+
+                if self._should_terminate:
+                    break
+                self._re_sync_alarms()
+
+            # Tell the terminating thread that it's safe to
+            # exit.
+            _log.info('Alarm manager shut down.')
+            self._condition.notify()
+
+    def terminate(self):
+        """Stop the run loop cleanly."""
+        _log.info('Shutting down alarm manager.')
+        with self._condition:
+            self._should_terminate = True
```

Comment:
I've since searched about 'with'. 
I just really dont like how it's layed out with the `self._blah`'s one space misaligned. 
But i guess this is just python layout.